src/libxtp/threecenter_gwbse.cc

```diff
@@ -161,5 +217,84 @@ void TCMatrix_gwbse::FillBlock(std::vector<Eigen::MatrixXd>& block,
   return;
 }
 
+/*
+ * Modify 3-center matrix elements consistent with use of symmetrized
+ * Coulomb interaction using OPENMP parallelization
+ */
+void TCMatrix_gwbse::MultiplyRightWithAuxMatrixOpenMP(
+    const Eigen::MatrixXd& matrix) {
+#pragma omp parallel for
+  for (int i_occ = 0; i_occ < _mtotal; i_occ++) {
+    Eigen::MatrixXd temp = _matrix[i_occ] * matrix;
+    _matrix[i_occ] = temp;
+  }
+  return;
+}
+
+#if defined(USE_CUDA)
+/*
+ * Convolution of the GW shell with the DFT orbital coefficients using an Nvidia
+ * GPU. The Cuda device behaves like a server that is receiving matrix-matrix
+ * multiplications from a single stream (an Nvidia queue) and handle them
+ * in an asynchronous way. It performs the following operations when recieving a
+ * request:
+ *  1. Check that there is enough space for the arrays
+ *  2. Allocate memory for each submatrix
+ *  3. copy the submatrix to the allocated space
+ *  4. return the result matrix
+ * The Cuda device knows to which memory address it needs to copy back the
+ * result. see: https://docs.nvidia.com/cuda/cublas/index.html#thread-safety2
+ * The amount of memory allocated in the device for this operations is roughly:
+ *  `3 * number_of_threads * (basisfunction ^ 2) + 2 * (basisfunction ^ 2)`
+ *  Meaning that if 3500 basis occupied around 0.1GB then 32 threads will
+ * allocate almost 10GB in the worst case scenario. Typical Nvidia devices
+ * have memory in the range between 10 to 48 GBs.
+ */
+void TCMatrix_gwbse::FillBlockCUDA(
+    std::vector<Eigen::MatrixXd>& block,
+    const std::vector<Eigen::MatrixXd>& symmstorage,
+    const std::pair<CudaMatrix, CudaMatrix>& cuda_matrices) {
+  EigenCuda cudaHandle;
+  int dim = static_cast<int>(symmstorage.size());
+  try {
+    for (int k = 0; k < dim; ++k) {
+      const Eigen::MatrixXd& matrix = symmstorage[k];
+      Eigen::MatrixXd threec_inMo = cudaHandle.triple_matrix_mult(
+          cuda_matrices.first, matrix.selfadjointView<Eigen::Lower>(),
+          cuda_matrices.second);
+      for (int i = 0; i < threec_inMo.cols(); ++i) {
+        block[i].col(k) = threec_inMo.col(i);
+      }
+    }
+  } catch (const std::runtime_error& error) {
+    XTP_LOG_SAVE(logDEBUG, _log)
+        << TimeStamp() << " FillBlockCUDA failed due to: " << error.what()
+        << " If there is not enough memory in the Nvidia GPU, try decreasing "
```

Comment:
this error message is not correct anymore

Reply:
What is this error not correct anymore? If there is not enough memory in the GPU the calculation is going to fail and the remedy is to lower the amount of GPU_STREAMS

Reply:
yes it says OPENMP threads though :D